feature_ShieldOverTime

Centorios opened the pull request:
-Introduces the RequireArmor field to t_Hechizo and updates spell loading and casting logic to support specific armor requirements. Refactors ProtectEffect to handle shield ticks, modifiers, and effect removal.
-Updates item requirement messaging and ensures user speed is refreshed when armor is unequipped.
-Added generic required item for this action locale msg
-Added Graphical and sound support to being hit and getting the effect removed
-Merge with https://github.com/ao-org/argentum-online-client/pull/691

Centorios committed:
Add RequireArmor property to spells and improve shield effect

Introduces the RequireArmor field to t_Hechizo and updates spell loading and casting logic to support specific armor requirements. Refactors ProtectEffect to handle shield ticks, modifiers, and effect removal.
Also updates item requirement messaging and ensures user speed is refreshed when armor is unequipped.

Centorios committed:
Added generic item required msg

ao-20 commented:
Fellow citizen of Nix, one of the admins verify this pull request

Centorios committed:
Merge branch 'master' into feature_ShieldOverTime

Centorios commented on Codigo/EffectOverTime/ProtectEffect.cls:
```diff
@@ -147,7 +166,13 @@ Public Sub IBaseEffectOverTime_OnRemove()
+        Call AddShieldToReference(DotInfo.TargetRef, -UserList(DotInfo.TargetRef.ArrayIndex).Stats.shield)
+    Else
+        Call AddShieldToReference(DotInfo.TargetRef, -NpcList(DotInfo.TargetRef.ArrayIndex).Stats.shield)
+    End If
```
i wanted to do this, but didn't know how.
I tried to track the specific ammount of shield generated by a determined instance of the buff with a private variable but it doesn't work, i need help with this, dont know how to track it with a vb6 variable

Centorios replied:
after struggling a lot to use native types, i implemented it with a dictionary, works flawlesly, and as intended

Centorios committed:
Add OnRemoveFx/Wav to EOT and improve shield handling

Introduces OnRemoveFx and OnRemoveWav fields to t_EffectOverTime and loads them in FileIO. Updates ProtectEffect to trigger these effects on removal and to play OnHitWav on hit. Refactors shield handling to use a unique accumulator per effect instance and ensures shield values do not go below zero in AddShieldToReference.

Centorios committed:
Fix FX and wave playback position for user effects

Corrects the position parameters for OnRemoveWav and OnHitWav sound playback to use the user's coordinates. Also ensures FX and wave effects are only sent for user references, improving consistency and preventing errors for non-user targets.

morgolock committed:
Merge branch 'master' into feature_ShieldOverTime

morgolock committed:
Merge branch 'master' into feature_ShieldOverTime

morgolock commented on Codigo/EffectOverTime/ProtectEffect.cls:
```diff
@@ -172,13 +210,18 @@ Public Function IBaseEffectOverTime_ApplyDamageReduction(ByVal Damage As Long, _
+                   DotInfo.EotId).OnHitFx, 0, UserList(DotInfo.TargetRef.ArrayIndex).pos.x, UserList(DotInfo.TargetRef.ArrayIndex).pos.y))
+            Else
+                Call SendData(SendTarget.ToIndex, DotInfo.TargetRef.ArrayIndex, PrepareMessageCreateFX(UserList(DotInfo.TargetRef.ArrayIndex).Char.charindex, EffectOverTime( _
+                   DotInfo.EotId).OnHitFx, 0, UserList(DotInfo.TargetRef.ArrayIndex).pos.x, UserList(DotInfo.TargetRef.ArrayIndex).pos.y))
```
En vez de 

```
Call SendData(SendTarget.ToPCAliveArea, DotInfo.TargetRef.ArrayIndex, PrepareMessageCreateFX(UserList(DotInfo.TargetRef.ArrayIndex).Char.charindex, EffectOverTime( _
                   DotInfo.EotId).OnHitFx, 0, UserList(DotInfo.TargetRef.ArrayIndex).pos.x, UserList(DotInfo.TargetRef.ArrayIndex).pos.y))
```

Usemos
```
with DotInfo
Call SendData(SendTarget.ToPCAliveArea, .TargetRef.ArrayIndex, PrepareMessageCreateFX(UserList(.TargetRef.ArrayIndex).Char.charindex, EffectOverTime( _
                   .EotId).OnHitFx, 0, UserList(.TargetRef.ArrayIndex).pos.x, UserList(.TargetRef.ArrayIndex).pos.y))
end with
```

Centorios committed:
Refactor shield effect removal and add error handling

Extracted user shield effect removal logic into a new RemoveUserShieldEffects subroutine for better code reuse. Updated damage reduction logic to use With DotInfo for clarity. Added error handling to PerformAction to trace errors. Adjusted shield calculation to use a constant percentage and improved skill cooldown update logic.

morgolock committed:
Merge branch 'master' into feature_ShieldOverTime

Centorios committed:
Merge branch 'master' into feature_ShieldOverTime

Centorios committed:
Use boolean semantics for DivineBlood flag

Replace numeric comparisons/assignments for the .flags.DivineBlood flag with boolean True/False and normalize checks across several modules for clarity and consistency. Files changed: Codigo/InvUsuario.bas, Codigo/Modulo_UsUaRiOs.bas, Codigo/TCP.bas, Codigo/UpdateHpOverTime.cls, Codigo/modHechizos.bas. Also removed a conditional guard in modHechizos.bas (the IsFeatureEnabled/DivineBlood/IsSet check) so the subsequent localized message and early return are now unconditional.

Centorios committed:
Update modHechizos.bas

Centorios committed:
Merge branch 'master' into feature_ShieldOverTime

Centorios committed:
Update GameLogic.bas

Centorios committed:
Update GameLogic.bas

Centorios committed:
Merge branch 'master' into feature_ShieldOverTime

Centorios committed:
Merge branch 'master' into feature_ShieldOverTime
